### ISSUE TITLE
Pause game on connection problems

### ISSUE THREAD

**Issue by DeinFreund** [ZeroK-RTS/Zero-K#1719]
Most payers don't seem to have an eye on the player list and even miss it when somebody requests a pause via lobby chat. An automated pause as soon as somebody quits, his ping grows over four seconds or he loses connection would reduce the impact of the still regularly occurring crashes. The pause should be overridable by players in case of a lagging person.

This should probably be limited to small teams, 1v1 and FFA to avoid grinding games to a hold.

**Comment by Licho1** [comment]
If its limited to small games/FFA its fine. For big games it would be too annoying and useless.

**Comment by GoogleFrog** [comment]
I don't think this is useful. The "wait or win" popup is pretty large so this is not an issue for 1v1. If the dropped player has a team then someone on the team will notice that they have received their dropped allies units.

**Comment by DeinFreund** [comment]
It is, most of the 1v1 where I dropped it took the opposing player 30-60 seconds to pause. Don't tell me you like to play 1v1 with random 1 minute drops.

Giving the enemy the option to win the game because the engine is unstable is even worse imo.. 

All of this just leads to us being in need of a non-development engine.

**Comment by GoogleFrog** [comment]
The "wait or win" popup appears as soon as you have definitely dropped so it's not like pause could detect it any earlier.

> Giving the enemy the option to win the game because the engine is unstable is even worse imo..

I don't know how you can think this. If your opponent drops you need to be able to say "they're gone, I'm not waiting for them".

**Comment by DeinFreund** [comment]
The current pop-up is a nice solution for now, so no need to implement something better currently, yet here's how I would expect it from a competitive game:

1. Player's ping goes above 5 seconds or player is inactive for 10 seconds
2. 1v1: game is paused - Teams: Units are given to teammates(already works) and there is a pop up informing everyone
3. The player has 5 minutes to rejoin the game
4. If the player fails to do so, the game is aborted (1v1, counts as loss) or continues without him (teams)
5. The player receives a temporary matchmaking ban

Matchmaking bans should exponentially increase in duration (but the counter decays over time) so people regularly dropping in competitive games would be excluded from them.

This is because even if the spring engine is perfect, computers still crash from time to time and internet has outages/disconnects.(Mostly GPU drivers in my case)

**Comment by Anarchid** [comment]
> The player has 5 minutes to rejoin the game

5 minutes is long. This will just lead to people aborting unilaterally rather than waiting, i think.

**Comment by DeinFreund** [comment]
Make it 3 minutes then. This is what is used in "a commercial game" and worked for me to even restart after bluescreens.

Keep in mind bans should range from days to week for repeat offenders. Also the game should not ever crash, so this won't ever happen ;)

**Comment by GoogleFrog** [comment]
My only other 1v1 online competitive experience is Starcraft II and I don't recall them pausing upon problems. This sounds like something that would be worth adding in some situations but not at the moment. Pausing and banning for competitive 1v1 makes the most sense when there are many people who care about improving and measuring their skill. In other situations it might just get in the way. I made sure that matchmaker banning was implemented because there should be a time when it is useful.

In the current situation I think it's not unreasonable to expect a manual courtesy pause and automatic banning will just extend the time it takes to try matchmaking again. In a situation of player influx it is more important to get players into games, and not block them, than it is to measure their skill accurately. Someone who is banned from matchmaking for a short period is likely to just give up entirely.

Also, I do not know how to implement an enforced pause. I doubt gadgets can even pause the game and there is probably a console command to toggle pause. I thought this ticket was about an automatic courtesy pause.

**Comment by db81** [comment]
> The "wait or win" popup appears as soon as you have definitely dropped so it's not like pause could detect it any earlier.

Isn't it possible to detect an ever-increasing ping or lack of cpu usage updates much sooner? At least in 1v1, I don't see the point of waiting for the rather long engine timeout. Just disable the win option until they're definitelly out.

**Comment by GoogleFrog** [comment]
Can gadgets pause the game?

**Comment by db81** [comment]
I think it's pretty unimportant to *enforce* pause. Just send a pause command from a widget when wait or win windows pops up and it's a 1v1. We can rely on a gentlemen's agreement to wait a reasonable amount of time before unpausing.

**Comment by sprunk** [comment]
As far as I know synced gadgets can `Spring.SendCommands("pause 1")` without problems and result in immediate pause (without network lag).

**Comment by sprunk** [comment]
This existed (927af3cbcee66c06d5f6101fd27d43d6b3c17c4d) and was reverted (d8a5ebd31abd87fc92af4e47c8c1dcc516baac70). Can probably still be brought back using more reliable detection?

**Comment by GoogleFrog** [comment]
I'm not against pausing on the connection popup screen. It just has to work perfectly and not be confusing for players. Remember that players may not even know how to unpause the game. Perhaps the popup should appear more centrally alongside the pause.
